### ISSUE TITLE
CI with more distinct test categories

### ISSUE THREAD

**Issue by bertllll** [MASQ-Project/MASQ-Node-issues#771]
Regarding CI test segmentation, I know that we’ve been able to roll on with we have and been happy, accepting a workaround in the integration tests. I’m standing at the state with PaymentAdjuster (practically complete), though, where I’ve written quite an important occasional test that assures me the machinery has no obvious weak spots because this way I can burden it with any number of auto-generated situations and see if anything fragile in there cracks. 

The problem is we’ve never really adopted a system for using these time-expensive tests and therefore our workspace is not ready for us to go on. 

I inspected what would be needed and my opinion is it doesn’t have a standard solution constituted of anything we’ve known and used. That said, I do have a solution that I consider clean but can imagine the others might not go along with it. 

It is a foxy and little bit hard-to-catch-on approach. If I want to be able to select a subset of tests that should compile and filter out the rest I’d need to employ features combined with the standard frame of conditional compilation (cfg).

We could have three sorts of tests:
feature=“integration_test”
feature=“occasional_test”
…and the third one that doesn’t have any of these features above, therefore, by default, as well as we have it now, those would be the most frequent ones, unit tests. 

However it’s not simple like that. Because the third group doesn’t have its own feature label, we can never disable it. It would always contaminate the other groups. 

Naturally, the next precaution seems to be addition of a third label. But now let’s try imagining how many tests would have to wear this somewhat awkward extra attribute, and second, it would throw more things to do and repeat all the time at the developers. Jo

Finally arriving at the design being proposed by me, I think we would need an attribute-like procedure macro. Let’s say it would look like this. #[masq_test] It would take care of the dirty work for us. It would allow these cases:

#[masq_test(integration)] 
#[masq_test(occasional)]
#[masq_test]

where the last one stands for a standard unit test. 

This convenient macro would, however, do the following  under the hood by expanding as:

#[test]
#[cfg(feature=“integration”)]

#[test]
#[cfg(feature=“occasional”)]

#[test] 
#[cfg(feature=“unit”)]

through which we would bypass the tedious and forget-prone manual editing.

This might be a pretty tiny card if played actually, advocated for taking place before the PaymentAdjuster card itself. Technically, once the procedure macro exists, I would use automated tools (with help of regex) to replace the old traditional #[test] with #[masq_test] (or #[mtest] or whatever we think up together) and just after that I would fix those few integration tests with the right new attribute. 

Basically, I’m asking for support of this plan or dismissal. If you can see my point then please make this card a priority card in my line of tasks. 

A serious weakness of this plan I know of is that anybody who would violate this new order by using the old #[test] would cause some mess without any raise of alarm or anything, the mess would stay with us until somebody accidentally realised what what was going on.

**Comment by kauri-hero** [comment]
Consider this approach to allow testing to be designated by category and also on user-defined situations:
eg. smoke tests

Can consider different approaches for the testing suite itself for more customization

**Comment by dnwiebe** [comment]
For what it's worth, I have no objection to the proposal above.

However, @kauri-hero 's suggestion about smoke tests might be a little problematic. Usually, a smoke test is written in a scripting language like bash, rather than a programming language like Rust, and therefore bringing @bertllll 's macro to bear on it would be fairly clumsy.

For example, here's what a smoke test for Node might look like (assuming that a release has been built and tagged):

Run a bash script that does the following:
1. Start the Daemon
2. Using the non-interactive mode of `masq`:
	a. Use the setup command to prepare the Node to connect into the existing network
	b. Start the Node
	c. Enter a loop to periodically check the Node's connection status
	d. If a timeout is reached, fail
	e. When the Node is fully connected, shut it down
3. Kill the Daemon

This _could_ be done in Rust, but it'd be a lot more tedious.

**Comment by bertllll** [comment]
Regardless what comes out from the discussion about bringing in a new category of ‘Smoke tests’, I and Dan have agreed that there is no factual blockade for deploying a category of ‘Ocassional tests’ with or without the first mentioned category in parallel. 

Therefore I have a tendency towards saying the card has a green sign for us devs. The decision goes to @kauri-hero to let us proceed and prioritise this (clearly not bloating threatened) smaller card or push it away for lack of interest now. I’m saying I’d like to have this test feature on hand in the future, but even now when one of my test actually matches this new category. (If this card is not freed I’ll have to add the teat #[ignore] attribute as it seems to me like the smoothest provisional solution of the situation until we establish the categories. I hope we will remember this test to be added when we have it.
